### PR TITLE
Add Transform class.

### PR DESCRIPTION
**Pull request by 1chandu**
- Adding new class, `Transform` easy interface to perform TransformFeedback iterations.
- Add unit tests and update TF examples to use new API.
- Add Documentation with use cases.

Verified with TF examples and Wind Demo (deck.gl).

### CODE REVIEW

File: src/core/experimental/buffer-map.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    let fs = PASS_THROUGH_FS;
+    const vsLines = vs.split('\n');
+    if (vsLines[0].indexOf('#version ') === 0) {
+      fs = `\
+${vsLines[0]}
+${PASS_THROUGH_FS}
+`;
+    }
+
+    this.model = new Model(this.gl, {
```
Comment (ibgreen) [inline src/core/experimental/buffer-map.js]: Passing through a `shaderCache` seems important? If not, does that mean we cannot use precompiled shader instances here?

File: src/core/experimental/buffer-map.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import Model from '../model';
+import {isWebGL2, assertWebGL2Context} from '../../webgl/context';
+import assert from 'assert';
+
+const PASS_THROUGH_FS = `\
+void main()
+{
+}`;
+
+export default class BufferMap {
```
Comment (ibgreen) [inline src/core/experimental/buffer-map.js]: I propose renaming to `Transform` as the official name for this class. Everyone talks about transform feedback and it relates to that, while sounding simpler (which is the purpose of the class).

File: examples/core/transform-feedback-instanced/app.js [hunk 3]
```diff
@@ -271,25 +234,34 @@ const animationLoop = new AnimationLoop({
...
-      transformFeedback: transformFeedbacks[destinationIdx]
-    });
+    bufferMap.run();
+
+    const offsetBuffer = bufferMap.getBuffer('v_offset');
+    const rotationBuffer = bufferMap.getBuffer('v_rotation');
 
+    bufferMap.swapBuffers();
+
+    offsetBuffer.updateLayout({instanced: 1});
```
Comment (ibgreen) [inline examples/core/transform-feedback-instanced/app.js]: opportunity for further improvement:

Having to do these layout updates really complicates the API. I think there is a way to avoid this, let's discuss.

File: examples/core/transform-feedback-instanced/app.js [hunk 4]
```diff
@@ -271,25 +234,34 @@ const animationLoop = new AnimationLoop({
...
-    transformVertexArrays,
-    transformFeedbacks,
     modelRender,
-    modelTransform
+    positionBuffer,
+    colorBuffer,
+    bufferMap
   }) {
 
     gl.viewport(0, 0, width, height);
```
Comment (ibgreen) [inline examples/core/transform-feedback-instanced/app.js]: I think gl.viewport is already handled by AnimationLoop

File: examples/core/transform-feedback-instanced/app.js [hunk 5]
```diff
@@ -176,80 +158,60 @@ const animationLoop = new AnimationLoop({
...
+      }
+    });
+
+    const bufferMap = new BufferMap(gl, {
+      sourceBuffers: {
+        a_offset: offsetBuffer,
+        a_rotation: rotationBuffer
+      },
+      vs: EMIT_VS,
+      varyings: ['v_offset', 'v_rotation'],
```
Comment (ibgreen) [inline examples/core/transform-feedback-instanced/app.js]: To me it seems we put specify the varyings twice. More intelligent defaults maybe we could make the varying lines optional?

File: examples/core/transform-feedback-instanced/app.js [hunk 6]
```diff
@@ -176,80 +158,60 @@ const animationLoop = new AnimationLoop({
...
+      type: gl.FLOAT
+    });
+
+    /* eslint-disable camelcase  */
+    const modelRender = new Model(gl, {
+      id: 'Model-Render',
+      vs: DRAW_VS,
+      fs: DRAW_FS,
+      drawMode: gl.TRIANGLE_FAN,
+      vertexCount: 3,
```
Comment (ibgreen) [inline examples/core/transform-feedback-instanced/app.js]: It seems to me that Model could autodeduce vertexCount, isInstanced and instanceCount from buffers. That would be a separate PR but a quite nice feature I think.

File: docs/api-reference/core/experimental/buffer-map.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+### bindBuffers
+
+Sets up source and destination buffers. Creates destination buffers when not provided. Either `destinationBuffers` or `sourceDestinationBufferMap` must be provided.
+
+* `sourceBuffers` (`Object`) - key and value pairs, where key is the name of vertex shader attribute and value is the corresponding `Buffer` object.
+* `destinationBuffers` (`Object`, Optional) - key and value pairs, where key is the name of vertex shader attribute and value is the corresponding `Buffer` object.
+* `sourceDestinationMap` (`Object`, Optional) - key and value pairs, where key is a vertex shader attribute name and value is a vertex shader varying name.
+
+### delete
```
Comment (ibgreen) [inline docs/api-reference/core/experimental/buffer-map.md]: Move to just after constructor

File: examples/core/transform-feedback-instanced/app.js (hunk 5 above)
Comment (1chandu) [reply examples/core/transform-feedback-instanced/app.js]: We can deduce `varyings` from `sourceDestinationMap`, but it is optional. We can make `varyings` optional only in cases where `sourceDestination` is given.

File: src/core/experimental/buffer-map.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+    for (let i = 0; i < 2; i++) {
+      this.transformFeedbacks[i] = new TransformFeedback(this.gl, {
+        buffers: this.destinationBuffers[i],
+        varyingMap: this.varyingMap
+      });
+    }
+
+    // Append matching version string to FS.
+    let fs = PASS_THROUGH_FS;
```
[the remark this answers is not in the document]
Comment (1chandu) [reply src/core/experimental/buffer-map.js]: `Model` object takes shader Strings as input, at this point we don't have access to `Shader` unless we create them and pass it to Model instead of Strings. But I like the idea of managing `version` string inside Shader, lets explore this idea in a separate PR.

File: src/core/experimental/buffer-map.js (hunk 1 above)
Comment (1chandu) [reply src/core/experimental/buffer-map.js]: This happens only once when `BufferMap` instance is constructed, but yes we can use precompiled shader instances of passthrough shader cache (one for each `version`), this falls into same set of changes as above comment, first need some changes in `Shader` to get version details, so next PR?

**Comment by 1chandu** [comment]
@ibgreen, create this to track remaining issues : https://github.com/uber/luma.gl/issues/401
Removed example/demo update from this PR, will open a new one.
